File: go-side/internal/layers/transmission/transmission.go
package transmission

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"strings"
	"sync"
)

const (
	Version      = byte(1)
	HeaderSize   = 16
	MaxFrameBits = 8 * 1024 * 1024
)

var magic = [4]byte{'C', 'C', '6', '7'}

var algorithmIDs = map[string]byte{"hamming": 1, "crc32": 2}
var idAlgorithms = map[byte]string{1: "hamming", 2: "crc32"}

type ReceivedFrame struct {
	Algorithm        string
	MessageBitLength int
	FrameBits        string
}

type Layer struct {
	connection net.Conn
	sendMutex  sync.Mutex
}

func New(connection net.Conn) *Layer {
	return &Layer{connection: connection}
}

func packBits(bits string) ([]byte, error) {
	for _, bit := range bits {
		if bit != '0' && bit != '1' {
			return nil, fmt.Errorf("la cadena solo puede contener bits 0 y 1")
		}
	}
	padded := bits + strings.Repeat("0", (8-len(bits)%8)%8)
	result := make([]byte, len(padded)/8)
	for index := range result {
		for _, bit := range padded[index*8 : index*8+8] {
			result[index] <<= 1
			result[index] |= byte(bit - '0')
		}
	}
	return result, nil
}

func unpackBits(data []byte, bitLength int) (string, error) {
	if bitLength > len(data)*8 {
		return "", fmt.Errorf("el cuerpo no contiene suficientes bits")
	}
	var builder strings.Builder
	builder.Grow(len(data) * 8)
	for _, value := range data {
		builder.WriteString(fmt.Sprintf("%08b", value))
	}
	return builder.String()[:bitLength], nil
}

func (layer *Layer) EnviarInformacion(algorithm string, messageBitLength int, frameBits string) error {
	algorithmID, ok := algorithmIDs[algorithm]
	if !ok {
		return fmt.Errorf("algoritmo no soportado")
	}
	if messageBitLength < 0 || uint64(messageBitLength) > uint64(^uint32(0)) {
		return fmt.Errorf("longitud de mensaje fuera de rango")
	}
	if len(frameBits) > MaxFrameBits {
		return fmt.Errorf("trama demasiado grande")
	}
	body, err := packBits(frameBits)
	if err != nil {
		return err
	}
	header := make([]byte, HeaderSize)
	copy(header[0:4], magic[:])
	header[4] = Version
	header[5] = algorithmID
	binary.BigEndian.PutUint16(header[6:8], 0)
	binary.BigEndian.PutUint32(header[8:12], uint32(messageBitLength))
	binary.BigEndian.PutUint32(header[12:16], uint32(len(frameBits)))

	layer.sendMutex.Lock()
	defer layer.sendMutex.Unlock()
	packet := append(header, body...)
	for len(packet) > 0 {
		written, writeErr := layer.connection.Write(packet)
		if writeErr != nil {
			return writeErr
		}
		if written == 0 {
			return io.ErrUnexpectedEOF
		}
		packet = packet[written:]
	}
	return nil
}

func (layer *Layer) RecibirInformacion() (*ReceivedFrame, error) {
	header := make([]byte, HeaderSize)
	if _, err := io.ReadFull(layer.connection, header); err != nil {
		if err == io.EOF {
			return nil, nil
		}
		return nil, err
	}
	if string(header[0:4]) != string(magic[:]) || header[4] != Version || binary.BigEndian.Uint16(header[6:8]) != 0 {
		return nil, fmt.Errorf("encabezado de transmision invalido")
	}
	algorithm, ok := idAlgorithms[header[5]]
	if !ok {
		return nil, fmt.Errorf("identificador de algoritmo desconocido")
	}
	messageLength := int(binary.BigEndian.Uint32(header[8:12]))
	frameLength := int(binary.BigEndian.Uint32(header[12:16]))
	if frameLength > MaxFrameBits {
		return nil, fmt.Errorf("trama recibida demasiado grande")
	}
	body := make([]byte, (frameLength+7)/8)
	if _, err := io.ReadFull(layer.connection, body); err != nil {
		return nil, err
	}
	bits, err := unpackBits(body, frameLength)
	if err != nil {
		return nil, err
	}
	return &ReceivedFrame{algorithm, messageLength, bits}, nil
}

File: python-side/atm/layers/transmission.py
"""Capa de Transmision: framing binario y E/S TCP."""

from dataclasses import dataclass
import socket
import struct
import threading

MAGIC = b"CC67"
VERSION = 1
HEADER = struct.Struct("!4sBBHII")
MAX_FRAME_BITS = 8 * 1024 * 1024
ALGORITHM_IDS = {"hamming": 1, "crc32": 2}
ID_ALGORITHMS = {value: key for key, value in ALGORITHM_IDS.items()}


@dataclass(frozen=True)
class ReceivedFrame:
    algorithm: str
    message_bit_length: int
    frame_bits: str


def pack_bits(bits: str) -> bytes:
    if any(bit not in "01" for bit in bits):
        raise ValueError("la cadena solo puede contener bits 0 y 1")
    padded = bits + "0" * ((8 - len(bits) % 8) % 8)
    return bytes(int(padded[i : i + 8], 2) for i in range(0, len(padded), 8))


def unpack_bits(data: bytes, bit_length: int) -> str:
    bits = "".join(f"{byte:08b}" for byte in data)
    if bit_length > len(bits):
        raise ValueError("el cuerpo no contiene suficientes bits")
    return bits[:bit_length]


def _recv_exact(sock: socket.socket, length: int) -> bytes | None:
    chunks = bytearray()
    while len(chunks) < length:
        chunk = sock.recv(length - len(chunks))
        if not chunk:
            if not chunks:
                return None
            raise ConnectionError("la conexion termino en medio de una trama")
        chunks.extend(chunk)
    return bytes(chunks)


class TransmissionLayer:
    def __init__(self, sock: socket.socket):
        self._sock = sock
        self._send_lock = threading.Lock()

    def enviar_informacion(self, algorithm: str, message_bit_length: int, frame_bits: str) -> None:
        if algorithm not in ALGORITHM_IDS:
            raise ValueError("algoritmo no soportado")
        if not 0 <= message_bit_length <= 0xFFFFFFFF:
            raise ValueError("longitud de mensaje fuera de rango")
        if len(frame_bits) > MAX_FRAME_BITS:
            raise ValueError("trama demasiado grande")
        header = HEADER.pack(
            MAGIC, VERSION, ALGORITHM_IDS[algorithm], 0,
            message_bit_length, len(frame_bits),
        )
        with self._send_lock:
            self._sock.sendall(header + pack_bits(frame_bits))

    def recibir_informacion(self) -> ReceivedFrame | None:
        raw_header = _recv_exact(self._sock, HEADER.size)
        if raw_header is None:
            return None
        magic, version, algorithm_id, flags, message_length, frame_length = HEADER.unpack(raw_header)
        if magic != MAGIC or version != VERSION or flags != 0:
            raise ValueError("encabezado de transmision invalido")
        if algorithm_id not in ID_ALGORITHMS:
            raise ValueError("identificador de algoritmo desconocido")
        if frame_length > MAX_FRAME_BITS:
            raise ValueError("trama recibida demasiado grande")
        body_length = (frame_length + 7) // 8
        body = _recv_exact(self._sock, body_length)
        if body is None:
            raise ConnectionError("falta el cuerpo de la trama")
        return ReceivedFrame(ID_ALGORITHMS[algorithm_id], message_length, unpack_bits(body, frame_length))
